### PR TITLE
fix: Search/Query may failed during updating delegator cache.

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #37115

casue init query node client is too heavy, so we remove updateShardClient from leader mutex, which cause much more concurrent cornor cases.

This PR delay query node client's init operation until `getClient` is called, then use leader mutex to protect updating shard client progress to avoid concurrent issues.

### CODE REVIEW

File: internal/proxy/meta_cache.go [hunk 1]
```diff
@@ -1016,6 +1007,15 @@ func (m *MetaCache) GetShards(ctx context.Context, withCache bool, database, col
 	// and create new client for new leaders
 	_ = m.shardMgr.UpdateShardLeaders(oldLeaders, ret)
```
Comment (xiaofan-luan) [inline internal/proxy/meta_cache.go:1008]: please remove the UpdateShardLeaders function and init the shartClient in getClient

**Comment by weiliu1031** [comment]
rerun ut

File: internal/proxy/shard_client.go [hunk 2]
```diff
@@ -31,26 +31,37 @@ var errClosed = errors.New("client is closed")
...
+	initialized atomic.Bool
+	creator     queryNodeCreatorFunc
 }
 
 func (n *shardClient) getClient(ctx context.Context) (types.QueryNodeClient, error) {
+	if !n.initialized.Load() {
+		n.Lock()
+		if !n.initialized.Load() {
+			if err := n.initClients(); err != nil {
+				return nil, err
```
Comment (xiaofan-luan) [inline internal/proxy/shard_client.go:50]: release n.lock

File: internal/proxy/meta_cache.go (hunk 1 above)
Comment (weiliu1031) [reply internal/proxy/meta_cache.go:1016]: refine the shard cache in next pr

**Comment by weiliu1031** [comment]
rerun go-sdk

**Comment by weiliu1031** [comment]
rerun ut

**Comment by weiliu1031** [comment]
rerun ut

**Pull request merged**